File: kotlin-remix-run-router/karakum/convertUtilityTypes.js
const ts = require("typescript")

module.exports = function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Partial"
            || node.typeName.text === "Readonly"
            || node.typeName.text === "Exclude"
        )
    ) {
        return render(node.typeArguments[0])
    }
    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertUtilityTypes.kt
package electron.karakum.plugins

import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.asArray
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertUtilityTypes = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Partial")

        val typeArguments = ensureNotNull(node.typeArguments?.asArray())

        render(typeArguments.first())
    }
}
